### ISSUE TITLE
[feature request] making pytorch less noisy 

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#68768]
## 🚀 Feature

How can we make pytorch be noise-conscious when it prints warnings.

For some reason when developers write code with warnings they at times miss thinking about the big picture and the deployment of the software.

## 1. repetitive multi-node logs

One of the big issues is that on multi-node the same warning gets printed as many times as there are processes. Which is a huge issue when one has say 512 processes.

Currently with 512 gpus I get this printed 512 times:

```
                               !! WARNING !!

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Your compiler (c++) is not compatible with the compiler Pytorch was
built with for this platform, which is g++ on linux. Please
use g++ to to compile your extension. Alternatively, you may
compile PyTorch from source using c++, and then you can also use
c++ to compile your extension.

See https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md for help
with compiling PyTorch from source.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

                              !! WARNING !!

  warnings.warn(WRONG_COMPILER_WARNING.format(
/gpfswork/rech/six/commun/conda/tr10-13B/lib/python3.8/site-packages/torch/utils/cpp_extension.py:295: UserWarning: 
```

How can we make pytorch be aware of uses in mutli-gpu environment and only print any such advisory warnings on global rank 0?

This is not the only warning, it's just the one I'm dealing with at this moment.

I used a workaround in the application where I shutdown warnings for the non-rank 0:

```
    if torch.distributed.get_rank() == 0:
        fused_kernels.load(args)
        torch.distributed.barrier()
    else:
        torch.distributed.barrier()
        with warnings.catch_warnings():
            # ignore loading noise
            warnings.simplefilter("ignore")
            fused_kernels.load(args)
```

Now it prints this on rank 0 only:

## 2. repetitive warnings in the same process

But since Megatron-LM builds a handful of these cuda extensions one after another this warnings still gets printed multiple times.

How can we make the software be aware that it already printed the warnings?

## 3. the warning needs to be printed only when it's relevant. 

The Megatron-LM extensions build just fine despite the warning and yet the warning gets printed every time the application is started regardless of whether it is going to build the warnings or not. Surely, this warning is relevant for the actual building, but if `cpp_extension.load` can see the extension doesn't need to be built why emit the irrelevant warning?

## 4. there needs to be a way to override a warning

Finally, a warning is useful the first 5 times it appears and if the user chooses not to do something about it, either by choice or often because they don't have that choice, there needs to be a way to tell pytorch to no longer emit this warning.

## 5. shutting warnings off is a bad idea

The other approach is for the user to just turn pytorch warnings off, but then they may miss important warnings.

This approach can't be used in frameworks since we can't decide for the user what they want or not.

Thank you!

**Comment by jbschlosser** [comment]
Triage review for visibility & to identify the correct labels.

**Comment by stas00** [comment]
Here are some ideas to address just the multi-node/multi-process issue (first sub-issue):

Practically, the framework or the end-user application most like should be able to do for each of the components it's using:

```
        if args.rank == 0:
            if args.log_level is not None:
                set_verbosity_pytorch(args.log_level)
                set_verbosity_transfomers(args.log_level)
                set_verbosity_deepspeed(args.log_level)
        else:
            if args.log_level_replica is not None:
                set_verbosity_pytorch(args.log_level)
                set_verbosity_transfomers(args.log_level)
                set_verbosity_deepspeed(args.log_level)
```

and then the user can add this to the application cmd:
```
    --log-level info \
    --log-level-replica error \
```

to take care of this need.

Of course, `set_verbosity_deepspeed` is a sort of a wrapper that does:

```
        def set_verbosity_deepspeed(logging_level: str):
            if not args.deepspeed: return
            from deepspeed.utils import logger as ds_logger
            log_level = logging.log_levels[logging_level]
            ds_logger.setLevel(log_level)
```

So need something similar for pytorch.

So this takes care of the loggers.

-------------------

And then there remains the `warn` system to be overridden.

unlike logger, this is much more complex it seems. 

In the OP I have given this example:

```
    if torch.distributed.get_rank() == 0:
        fused_kernels.load(args)
        torch.distributed.barrier()
    else:
        torch.distributed.barrier()
        with warnings.catch_warnings():
            # ignore loading noise
            warnings.simplefilter("ignore")
            fused_kernels.load(args)

```

which is somewhat similar to the logger overrides, but this requires a handpicked sections to override.

But in either case of logger warning vs `warn.warning` it's critical to distinguish between:
1. we know this is going to be replicated on every replica, when giving advice
2. this warning is unique to this replica, since say this particular gpu is broken, but most of the time this would more likely to be an assert. I just can't think of a good example here at the moment. 


------------------

And finally for the pytorch launcher, which happens before user application, so the user application can't control it. Thus there needs to be a flag that let's user control of whether to repeat the launcher info/warn logs on each process or not. e.g.:

```
python -u -m torch.distributed.run --logger_only_on_rank_0
```

**Comment by b0noI** [comment]
should this be filed as multiple feature requests?, while this can be used as an umbrella , discussions per sub-issues should probably be separated.

**Comment by malfet** [comment]
@mruberry users hates warnings, and we sometimes call `warn` excessively
@malfet should we have a mechanism to properly emit them only once in a lifetime?
And agree with @b0noI we should split this into a generic issue(about excessive warnings) and specific warning (about compiler incompatibility)
Will take an AI to dig into compiler incompatibility issue

**Comment by stas00** [comment]
I tried to present the complete case and how the different manifestations of the same warnings correlate to each other at different stages and in different situations which can't be appreciated if split into separate issues.

But please feel free to split it up to as many separate Issues as you see fit.

Thank you very much for looking into it.

**Comment by stas00** [comment]
> Will take an AI to dig into compiler incompatibility issue

in which case the user should have a way to say - yes, thank you for the warning, but I can see it's unwarranted in my situation as the extension builds just fine - so please can we not emit it again on this particular setup? 

And the particular setup can be some sort of HASH value made form the same inputs that were used to trigger the warning in the first place?

**Comment by malfet** [comment]
> This is only a part of the log, the 2nd part is:
> 
> ```
> Detected CUDA files, patching ldflags
> Emitting ninja build file /gpfsssd/worksf/projects/rech/six/commun/code/tr10-13B/Megatron-DeepSpeed/megatron/fused_kernels/build/build.ninja...
> Building extension module scaled_upper_triang_masked_softmax_cuda...
> Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
> ninja: no work to do.
> Loading extension module scaled_upper_triang_masked_softmax_cuda...
> ```

This should only be printed if [`torch.utils.cpp_extension.load`](https://pytorch.org/docs/stable/cpp_extension.html) is  invoked with `verbose` set to True (which it is not, by default)

@stas00 can you share how you invoke this codepath?
Also, for the incompatible compiler warning, can you please run `which c++` and `c++ --version` on the system in question and share its output there?

**Comment by stas00** [comment]
> This should only be printed if torch.utils.cpp_extension.load is invoked with verbose set to True (which it is not, by default)

Indeed, they use verbose=True on rank0:
https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/fused_kernels/__init__.py#L55

One is resolved! Thank you!

> Also, for the incompatible compiler warning, can you please run which c++ and c++ --version on the system in question and share its output there?

```
which c++; c++ --version
/usr/bin/c++
c++ (GCC) 8.3.1 20191121 (Red Hat 8.3.1-5)
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by stas00** [comment]
If it helps I'm introducing the concept of advisory warnings in HF Transformers https://github.com/huggingface/transformers/pull/14669

so rather than turning all warnings off, the framework developers can choose which warnings are only useful the first few times they are seen and if the user chooses not to act on those because they don't want to or cannot they can choose to turn those off by setting an environment variable.

**Comment by ezyang** [comment]
I just want to echo my full support for solving the problems outlined at the top of this post.
